Nodejs/Product/Nodejs/SharedProject/CommonProjectNode.DiskMerger.cs

```diff
@@ -22,41 +25,56 @@ internal abstract partial class CommonProjectNode
         /// </summary>
         private sealed class DiskMerger
         {
-            private readonly Stack<DirState> remainingDirs = new Stack<DirState>();
+            private readonly ConcurrentStack<(string Name, HierarchyNode Parent)> remainingDirs = new ConcurrentStack<(string, HierarchyNode)>();
             private readonly CommonProjectNode project;
 
             public DiskMerger(CommonProjectNode project, HierarchyNode parent, string dir)
             {
                 this.project = project;
-                this.remainingDirs.Push(new DirState(dir, parent));
+                this.remainingDirs.Push((dir, parent));
             }
 
             /// <summary>
             /// Continues processing the merge request, performing a portion of the full merge possibly
             /// returning before the merge has completed.
-            /// 
-            /// Returns true if the merge needs to continue, or false if the merge has completed.
             /// </summary>
-            public bool ContinueMerge(bool hierarchyCreated)
+            /// <returns>Returns true if the merge needs to continue, or false if the merge has completed.</returns>
+            public async Task<bool> ContinueMergeAsync(bool hierarchyCreated)
             {
                 if (this.remainingDirs.Count == 0)
                 {
                     // all done
-                    this.project.BoldStartupItem();
+                    await this.InvokeOnUIThread(this.project.BoldStartupItem);
                     return false;
                 }
 
-                var dir = this.remainingDirs.Pop();
-                if (!Directory.Exists(dir.Name))
+                if (!this.remainingDirs.TryPop(out var dir) || !Directory.Exists(dir.Name))
                 {
                     return true;
                 }
 
+                return await this.InvokeOnUIThread(() => this.ContinueMergeAsyncWorker(dir, hierarchyCreated));
+            }
+
+            private async Task<bool> ContinueMergeAsyncWorker((string Name, HierarchyNode Parent) dir, bool hierarchyCreated)
```

Comment:
Why pass `dir` as a tuple?  It looks like it's only used for its members.

Reply:
convenience when passing as argument